live/live-root/usr/lib/dracut/modules.d/99agama-dud/agama-dud-apply.sh

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+
+[ -e /dracut-state.sh ] && . /dracut-state.sh
+
+. /lib/dracut-lib.sh
+. /lib/img-lib.sh
+
+DUD_DIR="$NEWROOT/run/agama/dud"
+AGAMA_CLI="$NEWROOT/usr/bin/agama"
+AGAMA_DUD_INFO="/tmp/agamadud.info"
+DUD_RPM_REPOSITORY="$NEWROOT/var/lib/agama/dud/repo"
+
+shopt -s nullglob
+
+# Apply all the updates.
+#
+# Read the URL of the updates from $AGAMA_DUD_INFO and process each one.
+apply_updates() {
+  local file
+  local dud_url
+  local dud_root
+  index=0
+
+  while read -r dud_url; do
+    mkdir -p "$DUD_DIR"
+    filename=${dud_url##*/}
+    file="${DUD_DIR}/${filename}"
+    # FIXME: use an index because two updates, coming from different places, can have the same name.
+    echo "Fetching a Driver Update Disk from $dud_url to ${file}"
+    if ! $AGAMA_CLI download "$dud_url" "${file}"; then
+      warn "Failed to fetch the Driver Update Disk"
+      exit 1
+    fi
+
+    # The $dir could be a temporary one created by each function.
+    dir="$DUD_DIR/dud_$(printf "%03d" $index)"
+    mkdir -p "$dir"
+
+    format=$("$NEWROOT/usr/bin/chroot" "$NEWROOT" file "${file##"$NEWROOT"}")
+    case "$format" in
+    *RPM*)
+      apply_rpm_update "$file" "$dir"
+      ;;
+
+    *)
+      apply_dud_update "$file" "$dir"
+      ;;
+    esac
+
+    ((index++))
+  done <$AGAMA_DUD_INFO
+
+  create_repo "$DUD_RPM_REPOSITORY"
+}
+
+# Applies an update from an RPM package
+#
+# It extracts the RPM content and adjust the alternative links.
+apply_rpm_update() {
+  file=$1
+  dir=$2
+
+  echo "Apply update from an RPM package"
+  unpack_rpm "$file" "$dir"
+  install_update "$dir"
+}
+
+# Applies an update from an RPM package
+#
+#   1. Copy the inst-sys updates to the $NEWROOT system.
+#   2. Update agamactl, agama-autoyast and agama-proxy-setup alternative links.
+#   3. Copy the packages to the $DUD_RPM_REPOSITORY.
+apply_dud_update() {
+  file=$1
+  dir=$2
+
+  echo "Apply update from a Driver Update Disk archive"
+  unpack_img "$file" "$dir"
+  # FIXME: do not ignore the dist (e.g., "tw" in "x86_64-tw").
+  arch=$(uname -m)
+  dud_root=$(echo "${dir}/linux/suse/${arch}"-*)
+  install_update "${dud_root}/inst-sys"
+  copy_packages "$dud_root" "$DUD_RPM_REPOSITORY"
```

Comment:
As more inst.dud are possible and the DUD repo dir is just a single one, it may theoretically conflict if the same name of a package is used. There might be more packages under the same name, but in different version. In such case the last dud wins. It's only theoretical, but possible.